File: packages/dotnet/UCASSignIn.Core/Models/AttendanceState.cs
namespace UCASSignIn.Core;

public static class CachePolicy
{
    public static readonly TimeSpan ReferenceLifetime = TimeSpan.FromDays(7);
    public static bool Fresh(DateTimeOffset? updated, DateTimeOffset now, TimeSpan? lifetime = null)
        => updated is { } at && now >= at && now - at < (lifetime ?? ReferenceLifetime);
}

public enum AttendanceStatus { Unknown, Unsigned, Signed }
public enum AttendanceSource { Cache, Daily, Detail, Submission }

// Weekly/cache data never enters this reducer. Request revisions reject stale reads first.
public sealed record AttendanceEvidence(AttendanceStatus Status, AttendanceSource Source, DateTimeOffset ObservedAt,
    DateTimeOffset? LastSuccessfulSignAt = null, bool PendingVerification = false,
    AttendanceSource? NegativeSource = null)
{
    public AttendanceEvidence Observe(AttendanceStatus status, AttendanceSource source, DateTimeOffset now)
    {
        if (source == AttendanceSource.Submission)
            return new(AttendanceStatus.Signed, source, now, now, true);
        if (status == AttendanceStatus.Unknown)
            return Status == AttendanceStatus.Signed ? this with { NegativeSource = null }
                : new(status, source, now, LastSuccessfulSignAt);
        if (status == AttendanceStatus.Signed || Status != AttendanceStatus.Signed)
            return new(status, source, now, LastSuccessfulSignAt);
        // Keep the first conflict. A later independent read from the same source
        // can confirm the correction; no timers or elapsed-time assumptions.
        if (NegativeSource == source)
            return new(AttendanceStatus.Unsigned, source, now, LastSuccessfulSignAt);
        return this with { PendingVerification = true, NegativeSource = source };
    }
}

public sealed record AttendanceEvidenceCache(int Version, string AccountId,
    Dictionary<string, AttendanceEvidence> States, HashSet<string> InvalidatedCourses)
{
    public const int CurrentVersion = 1;
}

public interface IAttendanceStateStore
{
    Task<AttendanceEvidenceCache?> LoadAttendanceStateAsync(string accountId, CancellationToken ct = default);
    Task SaveAttendanceStateAsync(AttendanceEvidenceCache state, CancellationToken ct = default);
}

File: apps/apple/UCASSignIn/Core/AttendanceState.swift
import Foundation

public enum CachePolicy {
    public static let referenceLifetime: TimeInterval = 7 * 24 * 60 * 60
    public static func isFresh(_ updated: Date?, at now: Date, lifetime: TimeInterval = referenceLifetime) -> Bool {
        guard let updated else { return false }
        return now >= updated && now.timeIntervalSince(updated) < lifetime
    }
}

public enum AttendanceStatus: String, Codable, Sendable { case unknown, unsigned, signed }
public enum AttendanceSource: String, Codable, Sendable { case cache, daily, detail, submission }

public struct AttendanceEvidence: Codable, Equatable, Sendable {
    public var status: AttendanceStatus
    public var source: AttendanceSource
    public var observedAt: Date
    public var lastSuccessfulSignAt: Date?
    public var pendingVerification = false
    public var negativeSource: AttendanceSource?

    public init(status: AttendanceStatus = .unknown, source: AttendanceSource = .cache, observedAt: Date) {
        self.status = status; self.source = source; self.observedAt = observedAt
    }
    public func observing(_ value: AttendanceStatus, source: AttendanceSource, now: Date) -> Self {
        var next = self
        if source == .submission {
            next = Self(status: .signed, source: source, observedAt: now)
            next.lastSuccessfulSignAt = now; next.pendingVerification = true
            return next
        }
        if value == .unknown {
            if status != .signed {
                next = Self(status: .unknown, source: source, observedAt: now)
                next.lastSuccessfulSignAt = lastSuccessfulSignAt
                return next
            }
            next.negativeSource = nil
            return next
        }
        if value == .signed || status != .signed {
            next = Self(status: value, source: source, observedAt: now)
            next.lastSuccessfulSignAt = lastSuccessfulSignAt
            return next
        }
        // Only a later independent read from the same source can confirm a correction.
        if negativeSource == source {
            next = Self(status: .unsigned, source: source, observedAt: now)
            next.lastSuccessfulSignAt = lastSuccessfulSignAt
            return next
        }
        next.pendingVerification = true
        next.negativeSource = source
        return next
    }
}

public struct AttendanceEvidenceCache: Codable {
    public var version = 1
    public let accountID: String
    public var states: [String: AttendanceEvidence]
    public var invalidatedCourses: Set<String>
    public init(accountID: String, states: [String: AttendanceEvidence], invalidatedCourses: Set<String>) {
        self.accountID = accountID; self.states = states; self.invalidatedCourses = invalidatedCourses
    }
}
